In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import datetime as dt
from datetime import *

In [ ]:
pd.set_option('display.colheader_justify', 'center')
pd.set_option('max_colwidth', 200)

In [ ]:
# dfy = pd.read_csv("/content/drive/MyDrive/Office/rnd/SSv4/data/y.csv", index_col=0)
dfx = pd.read_csv("/content/drive/MyDrive/Office/rnd/SSv4/data/all_data.csv")
dfx["date_key"] = pd.to_datetime(dfx["date_key"], format="%Y%m%d")

In [ ]:
# dfx2 = pd.read_excel("/content/drive/MyDrive/Office/rnd/SSv4/data/compsaa_file2_20200701_20201231.xlsx")
# dfx2["date_key"] = pd.to_datetime(dfx2["date_key"], format="%Y%m%d")

In [ ]:
dfx["date_key"] = pd.to_datetime(dfx["date_key"], format="%Y%m%d")
# dfy['date_time'] =  pd.to_datetime(dfy['date_time'], infer_datetime_format=True)

# v1

In [ ]:
#####################################

from pandas.core.arrays import datetimes
uniques = list(set(dfx['sku']))

skus_with_no_comment = []
total_rows_with_this_sku = []
rows_with_no_comment_under_this_sku = []
rows_with_comments_under_this_sku = []
percentage_of_no_comment = []
highest_date_with_this_sku = []
lowest_date_with_this_sku = []
max_date_with_nocom = []
min_date_with_nocom = []
last_obs_far = []
last_obs_near = []
obs_near_close_to_30days = []

def create_dataframe():

  df1 = pd.DataFrame(list(zip(skus_with_no_comment, total_rows_with_this_sku, rows_with_no_comment_under_this_sku, 
                              rows_with_comments_under_this_sku, percentage_of_no_comment, highest_date_with_this_sku, lowest_date_with_this_sku, 
                              max_date_with_nocom, min_date_with_nocom, 
                              last_obs_far, last_obs_near, obs_near_close_to_30days)), 
                     columns = ["skus_with_no_comment", "total_rows", "rows_with_no_comment", 
                                "rows_with_comments", "no_comment_percentage", "highest_date", "lowest_date", "max_date_with_nocom", "min_date_with_nocom", 
                                "last_obs_far", "last_obs_near", "obs_near_close_to_30days"])

  return df1



def find_data(sku, df):

  df_n = df.loc[df['sku'] == sku].sort_values('date_key')
  # date = pd.to_datetime("today").date()
  
  date = "2021-10-12"  # Format: YYYY-MM-DD
  date = datetime.strptime(date, '%Y-%m-%d').date()
  
  # if df_n['total_comment_num'].isna().sum() > 0 or df_n['total_comment_num'] == 0.0:
  if ((df_n['total_comment_num'].isna().sum() > 0) | (df_n['total_comment_num'] == 0.0)).any():

    skus_with_no_comment.append(sku)
    total_rows_with_this_sku.append(len(df_n))
    rows_with_no_comment = len(df_n[(df_n['total_comment_num'].isna()) | (df_n['total_comment_num'] == 0.0)])
    rows_with_no_comment_under_this_sku.append(rows_with_no_comment)
    rows_with_comments_under_this_sku.append(len(df_n) - rows_with_no_comment)
    percentage_of_no_comment.append((df_n['total_comment_num'].isna().sum() / len(df_n)) * 100)
    highest_date_with_this_sku.append(df_n.loc[dfx['sku'] == sku, 'date_key'].max().date())
    lowest_date_with_this_sku.append(df_n.loc[dfx['sku'] == sku, 'date_key'].min().date())
    max_date_with_nocom.append(df_n.loc[(dfx['sku'] == sku) & (dfx['total_comment_num'].isna()),  'date_key'].max().date())
    min_date_with_nocom.append(df_n.loc[(dfx['sku'] == sku) & (dfx['total_comment_num'].isna()),  'date_key'].min().date())

    last_obs_far_value = 0
    last_obs_near_value = 0
    obs_near_close_to_30days_value = 0

    """
    last_obs_far.append()
    last_obs_near.append()
    obs_near_close_to_30days.append()
    """

    # last_obs_day_diff = min(np.absolute((df_n['date_key'].dt.date - pd.to_datetime("today").date()).dt.days))
    last_obs_day_diff = min(np.absolute((df_n['date_key'].dt.date - date).dt.days))
    
    if (len(df_n) - rows_with_no_comment) > 1:
      if last_obs_day_diff > 45:
        last_obs_far_value = 1
      else:
        last_obs_near_value = 1
      
        # thirty_days_ago = datetime.today().date() - timedelta(days=30)
        thirty_days_ago = date - timedelta(days=30)

        last_obs_day_diff30 = min(np.absolute((df_n['date_key'].dt.date - thirty_days_ago).dt.days))

        if last_obs_day_diff30 < 5:
          obs_near_close_to_30days_value = 1

    last_obs_far.append(last_obs_far_value)
    last_obs_near.append(last_obs_near_value)
    obs_near_close_to_30days.append(obs_near_close_to_30days_value)
  

  return create_dataframe()


def get_data(i, f):

  skus_with_no_comment.clear()
  total_rows_with_this_sku.clear()
  rows_with_no_comment_under_this_sku.clear()
  rows_with_comments_under_this_sku.clear()
  percentage_of_no_comment.clear()
  highest_date_with_this_sku.clear()
  lowest_date_with_this_sku.clear()
  max_date_with_nocom.clear()
  min_date_with_nocom.clear()
  last_obs_far.clear()
  last_obs_near.clear()
  obs_near_close_to_30days.clear()
  

  for s in uniques[i:f]:
    df = find_data(s, dfx)

  return df


df_test = get_data(0, 500)
df_test

skus_with_no_comment  total_rows  rows_with_no_comment  \
0          72497496064          7                7            
1         100013965313         50               25            
2          26916683778         11               10            
3          57662242818         60               37            
4       10023432093697          1                1            
..                    ...         ...                   ...   
491        56839111570          5                5            
492        60549497747          4                4            
493        37403493268         16               15            
494        37403493269          7                7            
495        60549497748          4                4            

     rows_with_comments  no_comment_percentage highest_date lowest_date  \
0             0               100.000000        2020-10-02   2020-08-19   
1            25                50.000000        2022-02-28   2021-09-28   
2             1                90.909091        2020-06-25   2020-01-08   
3            23                61.666667        2020-08-21   2020-01-03   
4             0               100.000000        2020-11-07   2020-11-07   
..                  ...                    ...          ...         ...   
491           0               100.000000        2020-01-16   2020-01-02   
492           0               100.000000        2020-02-04   2020-01-08   
493           1                93.750000        2020-05-07   2020-01-23   
494           0               100.000000        2020-08-13   2020-01-03   
495           0               100.000000        2020-02-04   2020-01-08   

    max_date_with_nocom min_date_with_nocom  last_obs_far  last_obs_near  \
0        2020-10-02          2020-08-19            0             0         
1        2022-02-28          2021-09-28            0             1         
2        2020-06-25          2020-01-08            0             0         
3        2020-08-13          2020-01-03            1             0         
4        2020-11-07          2020-11-07            0             0         
..                  ...                 ...           ...            ...   
491      2020-01-16          2020-01-02            0             0         
492      2020-02-04          2020-01-08            0             0         
493      2020-05-06          2020-01-23            0             0         
494      2020-08-13          2020-01-03            0             0         
495      2020-02-04          2020-01-08            0             0         

     obs_near_close_to_30days  
0                0             
1                0             
2                0             
3                0             
4                0             
..                        ...  
491              0             
492              0             
493              0             
494              0             
495              0             

[496 rows x 12 columns]

|index|skus\_with\_no\_comment|total\_rows|rows\_with\_no\_comment|rows\_with\_comments|no\_comment\_percentage|highest\_date|lowest\_date|max\_date\_with\_nocom|min\_date\_with\_nocom|last\_obs\_far|last\_obs\_near|obs\_near\_close\_to\_30days|
|---|---|---|---|---|---|---|---|---|---|---|---|---|
|0|72497496064|7|7|0|100\.0|2020-10-02|2020-08-19|2020-10-02|2020-08-19|0|0|0|
|1|100013965313|50|25|25|50\.0|2022-02-28|2021-09-28|2022-02-28|2021-09-28|0|1|0|
|2|26916683778|11|10|1|90\.9090909090909|2020-06-25|2020-01-08|2020-06-25|2020-01-08|0|0|0|
|3|57662242818|60|37|23|61\.66666666666667|2020-08-21|2020-01-03|2020-08-13|2020-01-03|1|0|0|
|4|10023432093697|1|1|0|100\.0|2020-11-07|2020-11-07|2020-11-07|2020-11-07|0|0|0|
|5|60879273989|2|2|0|100\.0|2020-01-03|2020-01-02|2020-01-03|2020-01-02|0|0|0|
|6|10074718218|199|113|86|56\.78391959798995|2022-02-22|2020-01-02|2022-02-21|2020-01-02|0|1|0|
|7|68684349452|6|6|0|100\.0|2020-06-25|2020-06-10|2020-06-25|2020-06-10|0|0|0|
|8|70223921164|1|1|0|100\.0|2020-07-23|2020-07-23|2020-07-23|2020-07-23|0|0|0|
|9|71524417552|3|3|0|100\.0|2020-10-08|2020-07-15|2020-10-08|2020-07-15|0|0|0|
|10|10032823402513|16|16|0|62\.5|2021-08-03|2021-06-26|2021-07-29|2021-06-26|0|0|0|
|11|64831356946|89|53|36|59\.55056179775281|2021-08-04|2020-04-22|2021-08-04|2020-04-22|1|0|0|
|12|63543443473|31|20|11|64\.51612903225806|2021-06-26|2020-09-18|2021-06-26|2020-11-12|1|0|0|
|13|10032823402512|14|14|0|64\.28571428571429|2021-08-03|2021-06-26|2021-07-29|2021-06-26|0|0|0|
|14|100030480404|12|5|7|41\.66666666666667|2022-02-28|2022-01-16|2022-02-28|2022-01-29|1|0|0|
|15|71224524825|26|26|0|73\.07692307692307|2021-01-12|2020-07-09|2021-01-09|2020-07-09|0|0|0|
|16|28705554462|67|40|27|59\.70149253731343|2022-02-07|2020-10-08|2022-02-06|2020-10-08|1|0|0|
|17|10030991802402|70|40|30|57\.14285714285714|2022-02-28|2021-06-21|2022-02-28|2021-06-21|0|1|1|
|18|10035736084515|14|14|0|57\.14285714285714|2021-10-12|2021-08-28|2021-10-11|2021-08-28|0|0|0|
|19|10035736084516|16|16|0|56\.25|2021-10-12|2021-08-28|2021-10-11|2021-08-28|0|0|0|

In [ ]:
len(uniques)

152722

# v2

In [ ]:
#####################################

from pandas.core.arrays import datetimes

skus_with_no_comment = []
total_rows_with_this_sku = []
rows_with_no_comment_under_this_sku = []
rows_with_comments_under_this_sku = []
percentage_of_no_comment = []
highest_date_with_this_sku = []
lowest_date_with_this_sku = []
max_date_with_nocom = []
min_date_with_nocom = []
last_obs_far = []
last_obs_near = []
obs_near_close_to_30days = []
last_obs_day_diffs = []



bad_scene_ones = []
bad_scene_twos = []
bad_scene_threes = []
bad_scene_fours = []
bad_scene_fives = []


out = {}



def find_data(sku, df):

  df_n = df.loc[df['sku'] == sku].sort_values('date_key')
  # date = pd.to_datetime("today").date()
  
  date = "2021-10-12"  # Format: YYYY-MM-DD
  date = datetime.strptime(date, '%Y-%m-%d').date()
  
  # if df_n['total_comment_num'].isna().sum() > 0 or df_n['total_comment_num'] == 0.0:
  if ((df_n['total_comment_num'].isna().sum() > 0) | (df_n['total_comment_num'] == 0.0)).any():

    skus_with_no_comment.append(sku)
    
    total_rows_with_this_sku.append(len(df_n))
    
    rows_with_no_comment = len(df_n[(df_n['total_comment_num'].isna()) | (df_n['total_comment_num'] == 0.0)])
    rows_with_no_comment_under_this_sku.append(rows_with_no_comment)
    
    rows_with_comments = len(df_n) - rows_with_no_comment
    rows_with_comments_under_this_sku.append(rows_with_comments)
    
    percentage_of_no_comment.append((df_n['total_comment_num'].isna().sum() / len(df_n)) * 100)


    highest_date_with_this_sku.append(df_n['date_key'].max().date())
    lowest_date_with_this_sku.append(df_n['date_key'].min().date())
    
    max_date_with_nocom.append(df_n.loc[df['total_comment_num'].isna(), 'date_key'].max().date())
    min_date_with_nocom.append(df_n.loc[df['total_comment_num'].isna(), 'date_key'].min().date())


    last_obs_far_value = 0
    last_obs_near_value = 0
    obs_near_close_to_30days_value = 0


    # last_obs_day_diff = min(np.absolute((df_n['date_key'].dt.date - pd.to_datetime("today").date()).dt.days))
    last_obs_day_diff = min(np.absolute((df_n['date_key'].dt.date - date).dt.days))
    last_obs_day_diffs.append(last_obs_day_diff)
    
    if (len(df_n) - rows_with_no_comment) > 1:
      if last_obs_day_diff > 45:
        last_obs_far_value = 1 # 1= True
      else:
        last_obs_near_value = 1
      
        # thirty_days_ago = datetime.today().date() - timedelta(days=30)
        thirty_days_ago = date - timedelta(days=30)

        last_obs_day_diff30 = min(np.absolute((df_n['date_key'].dt.date - thirty_days_ago).dt.days))

        if last_obs_day_diff30 <= 5:
          obs_near_close_to_30days_value = 1

    last_obs_far.append(last_obs_far_value)
    last_obs_near.append(last_obs_near_value)
    obs_near_close_to_30days.append(obs_near_close_to_30days_value)


    out = {
           "sku": skus_with_no_comment,
           "total_rows": total_rows_with_this_sku,
           "rows_with_no_comment": rows_with_no_comment_under_this_sku,
           "rows_with_comments": rows_with_comments_under_this_sku,
           "no_comment_percentage": percentage_of_no_comment,
           "last_obs_day_diff": last_obs_day_diffs,
           "last_obs_fars": last_obs_far,
           "last_obs_nears": last_obs_near,
           "obs_near_close_to_30dayss": obs_near_close_to_30days,
           "highest_date": highest_date_with_this_sku,
           "lowest_date": lowest_date_with_this_sku,
           "max_date_nocom": max_date_with_nocom,
           "min_date_nocom": min_date_with_nocom,
          }



    # if rows_with_comments == 0:
    """
      get all the true valued rows with bad scenarios
    """

    # bad_scenes = {
    #               "sku": skus_with_no_comment,
    #               "bad_scene_one": bad_scene_ones,
    #               "bad_scene_two": bad_scene_twos,
    #               "bad_scene_three": bad_scene_threes,
    #               "bad_scene_four": bad_scene_fours,
    #               "bad_scene_five": bad_scene_fives
    #              }


  return pd.DataFrame(out).set_index(['sku'])


def get_data(i, f, df):

  uniques = list(set(df['sku']))

  for s in uniques[i:f]:
    dff = find_data(s, df)

  return dff

df_test = get_data(0, 200, dfx)
df_test

total_rows  rows_with_no_comment  rows_with_comments  \
sku                                                                    
72497496064          7                7                    0           
100013965313        50               25                   25           
26916683778         11               10                    1           
57662242818         60               37                   23           
10023432093697       1                1                    0           
...                    ...                   ...                 ...   
10030812234070      27               27                    0           
10030812234071      26               26                    0           
42216194392         62               35                   27           
10034319720793       9                9                    0           
42216194389         17                8                    9           

                no_comment_percentage  last_obs_day_diff  last_obs_fars  \
sku                                                                       
72497496064          100.000000               375               0         
100013965313          50.000000                 0               0         
26916683778           90.909091               474               0         
57662242818           61.666667               417               1         
10023432093697       100.000000               339               0         
...                               ...                ...            ...   
10030812234070        66.666667                70               0         
10030812234071        50.000000                70               0         
42216194392           56.451613                 0               0         
10034319720793        44.444444                49               0         
42216194389           47.058824               268               1         

                last_obs_nears  obs_near_close_to_30dayss highest_date  \
sku                                                                      
72497496064            0                    0              2020-10-02    
100013965313           1                    0              2022-02-28    
26916683778            0                    0              2020-06-25    
57662242818            0                    0              2020-08-21    
10023432093697         0                    0              2020-11-07    
...                        ...                        ...          ...   
10030812234070         0                    0              2021-08-03    
10030812234071         0                    0              2021-08-03    
42216194392            1                    0              2022-02-21    
10034319720793         0                    0              2021-08-24    
42216194389            0                    0              2021-01-17    

               lowest_date max_date_nocom min_date_nocom  
sku                                                       
72497496064     2020-08-19   2020-10-02     2020-08-19    
100013965313    2021-09-28   2022-02-28     2021-09-28    
26916683778     2020-01-08   2020-06-25     2020-01-08    
57662242818     2020-01-03   2020-08-13     2020-01-03    
10023432093697  2020-11-07   2020-11-07     2020-11-07    
...                    ...            ...            ...  
10030812234070  2021-05-15   2021-07-29     2021-05-15    
10030812234071  2021-05-15   2021-07-29     2021-05-15    
42216194392     2020-11-08   2022-02-20     2020-11-08    
10034319720793  2021-07-29   2021-08-22     2021-07-29    
42216194389     2020-10-31   2021-01-15     2020-10-31    

[200 rows x 12 columns]

In [ ]:
len(uniques)

152722

In [ ]:
uniques

[72497496064,
 100013965313,
 26916683778,
 57662242818,
 10023432093697,
 60879273989,
 10074718218,
 68684349452,
 70223921164,
 71524417552,
 10032823402513,
 64831356946,
 63543443473,
 10032823402512,
 100030480404,
 71224524825,
 28705554462,
 10030991802402,
 10035736084515,
 10035736084516,
 10035736084517,
 10035736084518,
 69428314149,
 10031492235304,
 10031492235305,
 60038053930,
 60038053931,
 60038053932,
 10031492235306,
 10041578749998,
 10031492235307,
 10040943837229,
 10041578749997,
 10041578749996,
 10040815622726,
 10031492235308,
 66297004077,
 10024937324598,
 10040815622727,
 10041578749999,
 10031187099698,
 10041549652017,
 10671095868,
 10043335639102,
 1952448575,
 1952448576,
 10043335639104,
 10043335639106,
 100010033219,
 10043335639103,
 100010033221,
 10043335639107,
 10022610796615,
 66688122952,
 70066896968,
 70066896970,
 70066896969,
 100021567560,
 69394497612,
 12571639889,
 16144400473,
 16144400474,
 69471567963,
 10026064019546,
 1002638143

In [ ]:
date2 = "2021-10-12"  # Format: YYYY-MM-DD
date2 = datetime.strptime(date2, '%Y-%m-%d').date()

dftry = dfx.loc[dfx["sku"] == 57662242818]
day_diff = min(np.absolute((dftry['date_key'].dt.date - date2).dt.days))
 
day_diff

417

In [ ]:
print(df_test['lowest_date'].max())

2022-01-16


In [ ]:
#####################################

from pandas.core.arrays import datetimes

skus_with_no_comment = []
total_rows_with_this_sku = []
rows_with_no_comment_under_this_sku = []
rows_with_comments_under_this_sku = []
percentage_of_no_comment = []
highest_date_with_this_sku = []
lowest_date_with_this_sku = []
max_date_with_nocom = []
min_date_with_nocom = []
last_obs_far = []
last_obs_near = []
obs_near_close_to_30days = []
last_obs_day_diffs = []



def create_dataframe():

  df1 = pd.DataFrame(list(zip(skus_with_no_comment, total_rows_with_this_sku, rows_with_no_comment_under_this_sku, 
                              rows_with_comments_under_this_sku, percentage_of_no_comment, 
                              last_obs_day_diffs, last_obs_far, last_obs_near, obs_near_close_to_30days,
                              highest_date_with_this_sku, lowest_date_with_this_sku, 
                              max_date_with_nocom, min_date_with_nocom)), 
                     columns = ["skus_with_no_comment", "total_rows", "rows_with_no_comment", "rows_with_comments", "no_comment_percentage", 
                                "last_obs_day_diff", "last_obs_far", "last_obs_near", "obs_near_close_to_30days", 
                                "highest_date", "lowest_date", 
                                "max_date_with_nocom", "min_date_with_nocom"])

  return df1




def find_data(sku, df):

  df_n = df.loc[df['sku'] == sku].sort_values('date_key')
  # date = pd.to_datetime("today").date()
  
  date = "2021-10-12"  # Format: YYYY-MM-DD
  date = datetime.strptime(date, '%Y-%m-%d').date()
  
  # if df_n['total_comment_num'].isna().sum() > 0 or df_n['total_comment_num'] == 0.0:
  if ((df_n['total_comment_num'].isna().sum() > 0) | (df_n['total_comment_num'] == 0.0)).any():

    skus_with_no_comment.append(sku)
    total_rows_with_this_sku.append(len(df_n))

    rows_with_no_comment = len(df_n[(df_n['total_comment_num'].isna()) | (df_n['total_comment_num'] == 0.0)])
    rows_with_no_comment_under_this_sku.append(rows_with_no_comment)

    rows_with_comments = len(df_n) - rows_with_no_comment
    rows_with_comments_under_this_sku.append(rows_with_comments)

    # percentage_of_no_comment.append((df_n['total_comment_num'].isna().sum() / len(df_n)) * 100)
    percentage_of_no_comment.append((rows_with_no_comment / len(df_n)) * 100)

    highest_date_with_this_sku.append(df_n['date_key'].max().date())
    lowest_date_with_this_sku.append(df_n['date_key'].min().date())
    max_date_with_nocom.append(df_n.loc[df_n['total_comment_num'].isna(), 'date_key'].max().date())
    min_date_with_nocom.append(df_n.loc[df_n['total_comment_num'].isna(), 'date_key'].min().date())

    last_obs_far_value = 0
    last_obs_near_value = 0
    obs_near_close_to_30days_value = 0

    """
    last_obs_far.append()
    last_obs_near.append()
    obs_near_close_to_30days.append()
    """

    # last_obs_day_diff = min(np.absolute((df_n['date_key'].dt.date - pd.to_datetime("today").date()).dt.days))
    last_obs_day_diff = min(np.absolute((df_n['date_key'].dt.date - date).dt.days))

    last_obs_day_diffs.append(last_obs_day_diff)
    
    if (len(df_n) - rows_with_no_comment) > 1:
      if last_obs_day_diff > 45:
        last_obs_far_value = 1
      else:
        last_obs_near_value = 1

      
        # thirty_days_ago = datetime.today().date() - timedelta(days=30)
        thirty_days_ago = date - timedelta(days=30)

        last_obs_day_diff30 = min(np.absolute((df_n['date_key'].dt.date - thirty_days_ago).dt.days))

        if last_obs_day_diff30 < 5:
          obs_near_close_to_30days_value = 1

    last_obs_far.append(last_obs_far_value)
    last_obs_near.append(last_obs_near_value)
    obs_near_close_to_30days.append(obs_near_close_to_30days_value)
  

  return create_dataframe()


def get_data(i, f, df):

  skus_with_no_comment.clear()
  total_rows_with_this_sku.clear()
  rows_with_no_comment_under_this_sku.clear()
  rows_with_comments_under_this_sku.clear()
  percentage_of_no_comment.clear()
  highest_date_with_this_sku.clear()
  lowest_date_with_this_sku.clear()
  max_date_with_nocom.clear()
  min_date_with_nocom.clear()
  last_obs_far.clear()
  last_obs_near.clear()
  obs_near_close_to_30days.clear()
  last_obs_day_diffs.clear()
  

  uniques = list(set(df['sku']))

  for s in uniques[i:f]:
    dff = find_data(s, df)

  return dff


df_test = get_data(0, 1000, dfx)
df_test

skus_with_no_comment  total_rows  rows_with_no_comment  \
0          72497496064           7               7            
1         100013965313          50              25            
2          26916683778          11              10            
3          57662242818          60              37            
4       10023432093697           1               1            
..                    ...         ...                   ...   
984        65605994239          61              39            
985     10020495034113         141              81            
986     10032913057539           9               9            
987     10032913057540           8               8            
988        52183697157          44              25            

     rows_with_comments  no_comment_percentage  last_obs_day_diff  \
0             0               100.000000               375          
1            25                50.000000                 0          
2             1                90.909091               474          
3            23                61.666667               417          
4             0               100.000000               339          
..                  ...                    ...                ...   
984          22                63.934426                 1          
985          60                57.446809                 0          
986           0               100.000000                69          
987           0               100.000000                69          
988          19                56.818182               495          

     last_obs_far  last_obs_near  obs_near_close_to_30days highest_date  \
0          0             0                    0             2020-10-02    
1          0             1                    0             2022-02-28    
2          0             0                    0             2020-06-25    
3          1             0                    0             2020-08-21    
4          0             0                    0             2020-11-07    
..            ...            ...                       ...          ...   
984        0             1                    0             2021-12-27    
985        0             1                    1             2022-02-28    
986        0             0                    0             2021-08-04    
987        0             0                    0             2021-08-04    
988        1             0                    0             2020-06-04    

    lowest_date max_date_with_nocom min_date_with_nocom  
0    2020-08-19      2020-10-02          2020-08-19      
1    2021-09-28      2022-02-28          2021-09-28      
2    2020-01-08      2020-06-25          2020-01-08      
3    2020-01-03      2020-08-13          2020-01-03      
4    2020-11-07      2020-11-07          2020-11-07      
..          ...                 ...                 ...  
984  2020-04-03      2021-12-27          2020-04-03      
985  2020-08-28      2022-02-28          2020-08-28      
986  2021-07-21      2021-08-04          2021-07-21      
987  2021-07-20      2021-08-04          2021-07-20      
988  2020-01-02      2020-06-03          2020-01-02      

[989 rows x 13 columns]

In [ ]:
#####################################
#####################################
#####################################
#####################################

from pandas.core.arrays import datetimes

skus_with_no_comment = []
total_rows_with_this_sku = []
rows_with_no_comment_under_this_sku = []
rows_with_comments_under_this_sku = []
percentage_of_no_comment = []
highest_date_with_this_sku = []
lowest_date_with_this_sku = []
max_date_with_nocom = []
min_date_with_nocom = []
last_obs_far = []
last_obs_near = []
obs_near_close_to_30days = []
last_obs_day_diffs = []


badscene_one_lst = []
badscene_two_lst = []
badscene_three_lst = []
badscene_four_lst = []
badscene_five_lst = []
badscene_skus = []


def create_dataframe():

  df1 = pd.DataFrame(list(zip(skus_with_no_comment, total_rows_with_this_sku, rows_with_no_comment_under_this_sku, 
                              rows_with_comments_under_this_sku, percentage_of_no_comment, 
                              last_obs_day_diffs, last_obs_far, last_obs_near, obs_near_close_to_30days,
                              highest_date_with_this_sku, lowest_date_with_this_sku, 
                              max_date_with_nocom, min_date_with_nocom)), 
                     columns = ["skus_with_no_comment", "total_rows", "rows_with_no_comment", "rows_with_comments", "no_comment_percentage", 
                                "last_obs_day_diff", "last_obs_far", "last_obs_near", "obs_near_close_to_30days", 
                                "highest_date", "lowest_date", 
                                "max_date_with_nocom", "min_date_with_nocom"])

  return df1


def create_badscene_df():

  df1 = pd.DataFrame(list(zip(badscene_one_lst, badscene_two_lst, badscene_three_lst, badscene_four_lst, badscene_five_lst)), 
                     columns = ["badscenario_one", "badscenario_two", "badscenario_three", "badscenario_four", "badscenario_five"])

  return df1



def find_data(sku, df):

  df_n = df.loc[df['sku'] == sku].sort_values('date_key')
  # date = pd.to_datetime("today").date()
  
  date = "2021-10-12"  # Format: YYYY-MM-DD
  date = datetime.strptime(date, '%Y-%m-%d').date()
  
  # if df_n['total_comment_num'].isna().sum() > 0 or df_n['total_comment_num'] == 0.0:
  if ((df_n['total_comment_num'].isna().sum() > 0) | (df_n['total_comment_num'] == 0.0)).any():

    skus_with_no_comment.append(sku)
    total_rows_with_this_sku.append(len(df_n))

    rows_with_no_comment = len(df_n[(df_n['total_comment_num'].isna()) | (df_n['total_comment_num'] == 0.0)])
    rows_with_no_comment_under_this_sku.append(rows_with_no_comment)

    rows_with_comments = len(df_n) - rows_with_no_comment
    rows_with_comments_under_this_sku.append(rows_with_comments)

    # percentage_of_no_comment.append((df_n['total_comment_num'].isna().sum() / len(df_n)) * 100)
    percentage_of_no_comment.append((rows_with_no_comment / len(df_n)) * 100)

    highest_date_with_this_sku.append(df_n['date_key'].max().date())
    lowest_date_with_this_sku.append(df_n['date_key'].min().date())
    max_date_with_nocom.append(df_n.loc[df_n['total_comment_num'].isna(), 'date_key'].max().date())
    min_date_with_nocom.append(df_n.loc[df_n['total_comment_num'].isna(), 'date_key'].min().date())

    last_obs_far_value = 0
    last_obs_near_value = 0
    obs_near_close_to_30days_value = 0

    """
    last_obs_far.append()
    last_obs_near.append()
    obs_near_close_to_30days.append()
    """

    # last_obs_day_diff = min(np.absolute((df_n['date_key'].dt.date - pd.to_datetime("today").date()).dt.days))
    last_obs_day_diff = min(np.absolute((df_n['date_key'].dt.date - date).dt.days))

    last_obs_day_diffs.append(last_obs_day_diff)
    
    if (len(df_n) - rows_with_no_comment) > 1:
      if last_obs_day_diff > 45:
        last_obs_far_value = 1
        badscene_three_lst.append(sku)
      else:
        last_obs_near_value = 1
        badscene_four_lst.append(sku)

      
        # thirty_days_ago = datetime.today().date() - timedelta(days=30)
        thirty_days_ago = date - timedelta(days=30)

        last_obs_day_diff30 = min(np.absolute((df_n['date_key'].dt.date - thirty_days_ago).dt.days))

        if last_obs_day_diff30 < 5:
          obs_near_close_to_30days_value = 1
          badscene_five_lst.append(sku)

    last_obs_far.append(last_obs_far_value)
    last_obs_near.append(last_obs_near_value)
    obs_near_close_to_30days.append(obs_near_close_to_30days_value)


    if rows_with_comments == 0:
      # badscene_skus.append(sku)
      badscene_one_lst.append(sku)

    if rows_with_comments == 1:
      # badscene_skus.append(sku)
      badscene_two_lst.append(sku)
  

  return create_dataframe(), create_badscene_df()


def get_data(i, f, df):

  skus_with_no_comment.clear()
  total_rows_with_this_sku.clear()
  rows_with_no_comment_under_this_sku.clear()
  rows_with_comments_under_this_sku.clear()
  percentage_of_no_comment.clear()
  highest_date_with_this_sku.clear()
  lowest_date_with_this_sku.clear()
  max_date_with_nocom.clear()
  min_date_with_nocom.clear()
  last_obs_far.clear()
  last_obs_near.clear()
  obs_near_close_to_30days.clear()
  last_obs_day_diffs.clear()
  

  uniques = list(set(df['sku']))

  for s in uniques[i:f]:
    dff = find_data(s, df)

  return dff


df_test, df_try = get_data(0, 1000, dfx)
# display(df_test)
display(df_try)

badscenario_one  badscenario_two  badscenario_three  badscenario_four  \
0      72497496064      26916683778       57662242818       100013965313    
1   10023432093697   10041549652017       64831356946        10074718218    
2      60879273989       1952448575       63543443473     10030991802402    
3      68684349452      70066896969      100030480404        10671095868    
4      70223921164      52015399013       28705554462       100010033219    
5      71524417552      57664077937       60038053930       100021567560    
6   10032823402513   10029034373330       60038053931        16144400473    
7   10032823402512      64532775212       60038053932        16144400474    
8      71224524825      17745314100    10043335639102     10026064019546    
9   10035736084515      28663874016        1952448576        11670913144    
10  10035736084516      66790621745      100010033221     10038326591616    
11  10035736084517      71902954047       70066896968     10038326591617    
12  10035736084518      69653496431       70066896970        16593715330    
13     69428314149      63854609115       12571639889     10038326591618    
14  10031492235304      69035885315    10026381433525       100028645502    
15  10031492235305      71975830326       52015399014     10030486126734    
16  10031492235306      14647821196       56214945893        46175092891    
17  10041578749998      37403493268    10035675791468        46175092892    
18  10031492235307   10029499941816       25565069435        46175092893    
19  10040943837229   10033775510557       70529581181        46175092894    
20  10041578749997      57444140068       70529581182     10026412408989    
21  10041578749996      47713879196    10033185161368     10026412408988    
22  10040815622726      55767205102       11065622691     10022816579754    
23  10031492235308   10041255003430       57672728747     10022816579753    
24     66297004077   10027740693817       25150095532        30128734404    
25  10024937324598       1762395536       25150095531     10028680216765    
26  10040815622727      26814449159        1988886711        36522164504    
27  10041578749999      36970694187        1988886712        52093518121    
28  10031187099698      66329249341        1988886713        55125475628    
29  10043335639104      54041511511        1988886714     10024763785521    
30  10043335639106       1966081687        1988886716     10024763785524    
31  10043335639103       1966081689        1988886717     10024763785523    
32  10043335639107       1966081692       54979199165        32277528882    
33  10022610796615      56205248250       32953598143        42216194392    

    badscenario_five  
0    10030991802402   
1       16144400473   
2       16144400474   
3       46175092891   
4       46175092892   
5       46175092893   
6       46175092894   
7    10025651667409   
8    10025651667408   
9       67806167562   
10      71902954048   
11      71902954049   
12     100016849512   
13     100016849534   
14      71968490327   
15   10034604145232   
16     100006888420   
17      48149300197   
18     100006888444   
19       1981285382   
20      41138258962   
21      41138258963   
22      41138258965   
23     100001645604   
24      10494149714   
25   10032539870386   
26           787623   
27           787633   
28      55143826610   
29           787648   
30      67804333249   
31           787656   
32   10025482323208   
33          7079196

# Finding the skus under bad scenarios

In [ ]:
#####################################

from pandas.core.arrays import datetimes

skus_with_no_comment = []
total_rows_with_this_sku = []
rows_with_no_comment_under_this_sku = []
rows_with_comments_under_this_sku = []
percentage_of_no_comment = []
highest_date_with_this_sku = []
lowest_date_with_this_sku = []
max_date_with_nocom = []
min_date_with_nocom = []
last_obs_far = []
last_obs_near = []
obs_near_close_to_30days = []
last_obs_day_diffs = []



def create_dataframe():

  df_create = pd.DataFrame(list(zip(skus_with_no_comment, total_rows_with_this_sku, rows_with_no_comment_under_this_sku, 
                              rows_with_comments_under_this_sku, percentage_of_no_comment, 
                              last_obs_day_diffs, last_obs_far, last_obs_near, obs_near_close_to_30days,
                              highest_date_with_this_sku, lowest_date_with_this_sku, 
                              max_date_with_nocom, min_date_with_nocom)), 
                     columns = ["sku", "total_rows", "rows_with_no_comment", "rows_with_comments", "no_comment_percentage", 
                                "last_obs_day_diff", "last_obs_far", "last_obs_near", "obs_near_close_to_30days", 
                                "highest_date", "lowest_date", 
                                "max_date_with_nocom", "min_date_with_nocom"])
  # df_create = df_create.set_index("sku")

  return df_create




def find_data(sku, df):

  df_n = df.loc[df['sku'] == sku].sort_values('date_key')
  # date = pd.to_datetime("today").date()
  
  date = "2021-10-12"  # Format: YYYY-MM-DD
  date = datetime.strptime(date, '%Y-%m-%d').date()
  
  # if df_n['total_comment_num'].isna().sum() > 0 or df_n['total_comment_num'] == 0.0:
  if ((df_n['total_comment_num'].isna().sum() > 0) | (df_n['total_comment_num'] == 0.0)).any():

    skus_with_no_comment.append(sku)
    total_rows_with_this_sku.append(len(df_n))

    rows_with_no_comment = len(df_n[(df_n['total_comment_num'].isna()) | (df_n['total_comment_num'] == 0.0)])
    rows_with_no_comment_under_this_sku.append(rows_with_no_comment)

    rows_with_comments = len(df_n) - rows_with_no_comment
    rows_with_comments_under_this_sku.append(rows_with_comments)

    # percentage_of_no_comment.append((df_n['total_comment_num'].isna().sum() / len(df_n)) * 100)
    percentage_of_no_comment.append((rows_with_no_comment / len(df_n)) * 100)

    highest_date_with_this_sku.append(df_n['date_key'].max().date())
    lowest_date_with_this_sku.append(df_n['date_key'].min().date())
    max_date_with_nocom.append(df_n.loc[df_n['total_comment_num'].isna(), 'date_key'].max().date())
    min_date_with_nocom.append(df_n.loc[df_n['total_comment_num'].isna(), 'date_key'].min().date())

    last_obs_far_value = 0
    last_obs_near_value = 0
    obs_near_close_to_30days_value = 0

    """
    last_obs_far.append()
    last_obs_near.append()
    obs_near_close_to_30days.append()
    """

    # last_obs_day_diff = min(np.absolute((df_n['date_key'].dt.date - pd.to_datetime("today").date()).dt.days))
    last_obs_day_diff = min(np.absolute((df_n['date_key'].dt.date - date).dt.days))

    last_obs_day_diffs.append(last_obs_day_diff)
    
    if (len(df_n) - rows_with_no_comment) > 1:
      if last_obs_day_diff > 45:
        last_obs_far_value = 1
      else:
        last_obs_near_value = 1

      
        # thirty_days_ago = datetime.today().date() - timedelta(days=30)
        thirty_days_ago = date - timedelta(days=30)

        last_obs_day_diff30 = min(np.absolute((df_n['date_key'].dt.date - thirty_days_ago).dt.days))

        if last_obs_day_diff30 < 5:
          obs_near_close_to_30days_value = 1

    last_obs_far.append(last_obs_far_value)
    last_obs_near.append(last_obs_near_value)
    obs_near_close_to_30days.append(obs_near_close_to_30days_value)
  

  return create_dataframe()


def get_data(i, f, df):

  skus_with_no_comment.clear()
  total_rows_with_this_sku.clear()
  rows_with_no_comment_under_this_sku.clear()
  rows_with_comments_under_this_sku.clear()
  percentage_of_no_comment.clear()
  highest_date_with_this_sku.clear()
  lowest_date_with_this_sku.clear()
  max_date_with_nocom.clear()
  min_date_with_nocom.clear()
  last_obs_far.clear()
  last_obs_near.clear()
  obs_near_close_to_30days.clear()
  last_obs_day_diffs.clear()
  

  uniques = list(set(df['sku']))

  for s in uniques[i:f]:
    df_get = find_data(s, df)

  return df_get


def all_bad_scenes(df):
  df_all = df.loc[(df['rows_with_comments'] == 0) | (df['rows_with_comments'] == 1) | (df['last_obs_far'] == 1) | (df['last_obs_near'] == 1)
                    | df['obs_near_close_to_30days'] == 1]

  return df_all


def bad_scene1(df):
  df_bad1 = df.loc[(df['rows_with_comments'] == 0)]
  df_bad1['sku'].to_csv('badscenario_1.csv', mode='a', index=False, header=False)


  return df_bad1


def bad_scene2(df):
  df_bad2 = df.loc[(df['rows_with_comments'] == 1)]
  df_bad2['sku'].to_csv('badscenario_2.csv', mode='a', index=False, header=False)


  return df_bad2


def bad_scene3(df):
  df_bad3 = df.loc[(df['last_obs_far'] == 1)]
  df_bad3['sku'].to_csv('badscenario_3.csv', mode='a', index=False, header=False)


  return df_bad3


def bad_scene4(df):
  df_bad4 = df.loc[df['last_obs_near'] == 1]
  df_bad4['sku'].to_csv('badscenario_4.csv', mode='a', index=False, header=False)


  return df_bad4


def bad_scene5(df):
  df_bad5 = df.loc[df['obs_near_close_to_30days'] == 1]
  df_bad5['sku'].to_csv('badscenario_5.csv', mode='a', index=False, header=False)


  return df_bad5

In [ ]:
df_test = get_data(150001, 152722, dfx)
print(len(df_test))

df1 = bad_scene1(df_test)
print(len(df1))

df2 = bad_scene2(df_test)
print(len(df2))

df3 = bad_scene3(df_test)
print(len(df3))

df4 = bad_scene4(df_test)
print(len(df4))

df5 = bad_scene5(df_test)
print(len(df5))

2700
1613
102
593
392
116


In [ ]:
df1 = bad_scene1(df_test)
print(len(df1))

565


In [ ]:
df2 = bad_scene2(df_test)
print(len(df2))

35


In [ ]:
df3 = bad_scene3(df_test)
print(len(df3))

207


In [ ]:
df4 = bad_scene4(df_test)
print(len(df4))

184


In [ ]:
df5 = bad_scene5(df_test)
print(len(df5))

56


In [ ]:
%cd /content/drive/MyDrive/Office/rnd/SSv4/bad_scenarios

/content/drive/MyDrive/Office/rnd/SSv4/bad_scenarios


In [ ]:
len(uniques)

152722